components/scream/src/physics/p3/p3_functions_calc_rime_density_impl.hpp

```diff
@@ -0,0 +1,83 @@
+#ifndef P3_FUNCTIONS_CALC_RIME_DENSITY_IMPL_HPP
+#define P3_FUNCTIONS_CALC_RIME_DENSITY_IMPL_HPP
+
+#include "p3_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 rime density calculation function.
+ * Clients should NOT #include this file, but include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::calc_rime_density(const Spack& t, const Spack& rhofaci,
+                    const Spack& f1pr02, const Spack& acn,
+                    const Spack& lamc, const Spack& mu_c,
+                    const Spack& qc_incld, const Spack& qccol,
+                    Spack& vtrmi1, Spack& rhorime_c)
+{
+  constexpr Scalar qsmall = C::QSMALL;
+  constexpr Scalar ZeroDegC = C::ZeroDegC;
+  constexpr Scalar bcn = C::bcn;
+
+  const auto qccol_not_small_and_t_freezing = (qccol >= qsmall) &&
+                                              (t < ZeroDegC);
+
+  // NOTE: applicable for cloud only; modify when rain is added back
+  if (qccol_not_small_and_t_freezing.any()) {
+    // Mass-weighted mean ice fallspeed
+    vtrmi1.set(qccol_not_small_and_t_freezing, f1pr02 * rhofaci);
+
+    // If qc_incld isn't small, compute the rime density.
+    const auto qccol_and_qc_not_small_and_t_freezing = (qccol >= qsmall) &&
+                                                       (qc_incld >= qsmall) &&
+                                                       (t < ZeroDegC);
+    if (qccol_and_qc_not_small_and_t_freezing.any()) {
+
+      // Droplet fall speed (using Stokes' formulation, with analytic soln).
+      Spack Vt_qc;
+      Vt_qc.set(qccol_and_qc_not_small_and_t_freezing,
```

Comment:
For temporaries, you needn't use the masked set.

Reply:
Thanks. I didn't even think about that.